### DISCUSSION TITLE
dbt integration - adding python downstream and `dagster_dbt_translator`

### DISCUSSION THREAD

**Discussion by Al-Moatasem** [dagster-io/dagster#18461] [Q&A] [answered]
Hello Dagster team,

I am using `dagster==1.5.2`, and I have a few questions related to dbt integration `dagster-dbt==0.21.2`

1. How to add a downstream asset after dbt assets are materialized, regardless of the status success/failed, here the downstream asset doesn't depend on a single asset key.
   - The case I have, I need after running `dbt run` through Dagster, to copy artifact files into a cloud storage
![image](https://github.com/dagster-io/dagster/assets/47496345/37d75ad3-a9ac-4c38-9c8e-11696149d5bc)

---

2. What is the best practice for creating dedicated job for triggering `dbt run` and another job for triggering `dbt snapshot`? here is my current setup
```py
from pathlib import Path
from typing import Any, Mapping

from dagster import AssetKey, EnvVar, OpExecutionContext
from dagster_dbt import (
    DagsterDbtTranslator,
    DbtCliInvocation,
    DbtCliResource,
    build_schedule_from_dbt_selection,
    dbt_assets,
)

DBT_MANIFEST_PATH = Path(EnvVar("DBT_MANIFEST_PATH").get_value()).resolve()


class CustomDagsterDbtTranslator(DagsterDbtTranslator):
    @classmethod
    def get_asset_key(cls, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
        # default: ['dbt_schema','dbt_model_name']
        asset_key: AssetKey = super().get_asset_key(
            dbt_resource_props=dbt_resource_props
        )

        if dbt_resource_props.get("resource_type") == "model":
            asset_key = asset_key.with_prefix("model")
        elif dbt_resource_props.get("resource_type") == "snapshot":
            asset_key = asset_key.with_prefix("snapshot")

        return asset_key


@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:model,fqn:*",
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_models(
    context: OpExecutionContext,
    dbt: DbtCliResource,
):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["run"], context=context)

    yield from dbt_run_invocation.stream()


@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:snapshot,fqn:*",
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_snapshots(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["snapshot"], context=context)

    yield from dbt_run_invocation.stream()


dbt_models_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_models],
    job_name="daily_midnight_all_dbt_models",
    cron_schedule="@daily",
)

dbt_snapshots_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_snapshots],
    job_name="daily_midnight_all_dbt_snapshots",
    cron_schedule="@daily",
)
```
Note: For both functions decoratored by `@dbt_assets` I am passing `dagster_dbt_translator=CustomDagsterDbtTranslator()` 
This setup gives me this (both jobs have the same set of dbt resources)
![image](https://github.com/dagster-io/dagster/assets/47496345/33f68c45-59e8-4574-9cc8-37bc35619257)

If I removed the `dagster_dbt_translator=CustomDagsterDbtTranslator()` from one of these functions
```py
@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:snapshot,fqn:*",
    # ⛔
    # dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_snapshots(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["snapshot"], context=context)

    yield from dbt_run_invocation.stream()
```
I get what I expect, each job has its dbt resources
![image](https://github.com/dagster-io/dagster/assets/47496345/0c100312-fdd5-4452-a27d-f17332aa4bb9)

---

3. in both, `@dbt_assets()` decorator and `DbtCliResource.cli()` method, we can pass `dagster_dbt_translator` parameter, do we get different results if we use one instead of the other?

**Comment by rexledesma** [comment, marked as answer]
I'll answer your questions one-by-one here:

> 1. How to add a downstream asset after dbt assets are materialized, regardless of the status success/failed, here the downstream asset doesn't depend on a single asset key.

Rather than adding a downstream asset, you can do this in the body of your `@dbt_assets` decorated function. You can add your invocation in a `try/finally` block, where the `finally` block uploads the artifact files into storage.

```python
@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:model,fqn:*",
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_models(
    context: OpExecutionContext,
    dbt: DbtCliResource,
):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["run"], context=context)

    try:
        yield from dbt_run_invocation.stream()
    finally:
        dbt_manifest_path = dbt_run_invocation.target_path.joinpath("manifest.json")
        run_results_path = dbt_run_invocation.target_path.joinpath("run_results.json")
        
        # Or, get the json
        dbt_manifest_json = dbt_run_invocation.get_artifact("manifest.json")
        run_results_json = dbt_run_invocation.get_artifact("run_results.json")
        
        # Upload to cloud storage
        ...

```

> 2. What is the best practice for creating dedicated job for triggering dbt run and another job for triggering dbt snapshot?

You need to pass a dbt selection string to `build_schedule_from_dbt_selection`. The selection string is not automatically passed down from the `dbt_assets` object you pass in (this is a known issue from https://github.com/dagster-io/dagster/issues/15936). 

```diff
dbt_models_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_models],
    job_name="daily_midnight_all_dbt_models",
    cron_schedule="@daily",
+   dbt_select="resource_type:model,fqn:*"
)

dbt_snapshots_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_snapshots],
    job_name="daily_midnight_all_dbt_snapshots",
    cron_schedule="@daily",
+   dbt_select="resource_type:snapshot,fqn:*"
)
```

> 3. In both, `@dbt_assets()` decorator and `DbtCliResource.cli()` method, we can pass `dagster_dbt_translator` parameter, do we get different results if we use one instead of the other?

If you don't pass in a `dagster_dbt_translator` to `DbtCliResource.cli()` when you use it in a `@dbt_assets` decorated function, it will inherit the `dagster_dbt_translator` that you passed in to `@dbt_assets`.

If you explicitly pass in a `dagster_dbt_translator` to `DbtCliResource.cli()`, it will use that instead.

`dagster_dbt_translator` is an argument to `DbtCliResource.cli()` in case you want to invoke it in a non-asset context (e.g. in an op)

**Comment by Al-Moatasem** [comment]
@rexledesma, Thanks for your response.
- For the first question 
  > How to add a downstream asset after dbt assets are materialized, regardless of the status success/failed, here the downstream asset doesn't depend on a single asset key.

  in my case (uploading artifact files to the cloud) your approach will help, I have a **follow-up question** for the sake of understanding how to work with Dagster, in a different case, if I need to add an asset to the graph that depends on all assets in `dbt_assets()`, how to do so?

- for the second question
  I have applied your suggestion by adding `dbt_select="resource_type:xx,fqn:*"` in `build_schedule_from_dbt_selection()`, however, the same issue still exists (using `dagster_dbt_translator` in `dbt_assets()` in both decorated functions `all_dbt_models` and `all_dbt_snapshots` will return the same graph

  ![image](https://github.com/dagster-io/dagster/assets/47496345/550b06f6-3816-434a-bc70-a92cf2539d46)

  and if I removed `dagster_dbt_translator` from the decorated function `all_dbt_snapshots`, I got the expected graph

  I have these changes in the latest version of `dagster==1.5.10` and `dagster-dbt==0.21.10`

```py

class CustomDagsterDbtTranslator(DagsterDbtTranslator):
    @classmethod
    def get_asset_key(cls, dbt_resource_props: Mapping[str, Any]) -> AssetKey:
        asset_key: AssetKey = super().get_asset_key(
            dbt_resource_props=dbt_resource_props
        )

        if dbt_resource_props.get("resource_type") == "model":
            asset_key = asset_key.with_prefix("model")
        elif dbt_resource_props.get("resource_type") == "snapshot":
            asset_key = asset_key.with_prefix("snapshot")

        return asset_key

@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:model,fqn:*",
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_models(
    context: OpExecutionContext,
    dbt: DbtCliResource,
):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["run"], context=context)

    yield from dbt_run_invocation.stream()

@dbt_assets(
    manifest=DBT_MANIFEST_PATH,
    select="resource_type:snapshot,fqn:*",
    dagster_dbt_translator=CustomDagsterDbtTranslator(),
)
def all_dbt_snapshots(context: OpExecutionContext, dbt: DbtCliResource):
    dbt_run_invocation: DbtCliInvocation = dbt.cli(["snapshot"], context=context)

    yield from dbt_run_invocation.stream()


dbt_models_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_models],
    job_name="daily_midnight_all_dbt_models",
    dbt_select="resource_type:model,fqn:*",
    cron_schedule="@daily",
)

dbt_snapshots_daily = build_schedule_from_dbt_selection(
    dbt_assets=[all_dbt_snapshots],
    job_name="daily_midnight_all_dbt_snapshots",
    dbt_select="resource_type:snapshot,fqn:*",
    cron_schedule="@daily",
)
```

**Reply by rexledesma** [reply]
> If I need to add an asset to the graph that depends on all assets in dbt_assets(), how to do so?

You can create an asset that depends on all the asset keys from your dbt models. E.g.

```python
from dagster import asset


@asset(deps=all_dbt_models.keys)
def my_asset():
    ...
```

As for your other follow up, I can't replicate your issue. With my suggestion, `daily_midnight_all_dbt_models` should have only the dbt models, and `daily_midnight_all_dbt_snapshots` should have only the dbt snapshots. I tried your code on the jaffle shop dbt project and it was successful.

**Reply by Al-Moatasem** [reply]
@rexledesma, Thank you for your support, I will try to figure out why it is not working as expected on my end, and if I figure out the reason, I will post it here as an update for this post.
